kof-operator/internal/controller/clusterdeployment_kof_cluster_regional_role.go

```diff
@@ -2,668 +2,171 @@ package controller
 		return err
 	}
 
 	utils.LogEvent(
 		r.ctx,
-		"PromxyServerGroupCreated",
-		"PromxyServerGroup is successfully created",
+		"ConfigMapCreated",
+		"Created regional cluster ConfigMap",
 		r.clusterDeployment,
 		nil,
-		"promxyServerGroupName", promxyServerGroup.Name,
+		"configMapName", cm.Name,
+		"configMapNamespace", cm.Namespace,
+		"configMapData", configData,
 	)
 
 	return nil
 }
 
-func (r *RegionalClusterRole) GetGrafanaDatasource() (*grafanav1beta1.GrafanaDatasource, error) {
-	grafanaDatasource := &grafanav1beta1.GrafanaDatasource{}
-	if err := r.client.Get(r.ctx, types.NamespacedName{
-		Name:      r.GetGrafanaDatasourceName(),
-		Namespace: r.clusterNamespace,
-	}, grafanaDatasource); err != nil {
-		if errors.IsNotFound(err) {
-			return nil, nil
-		}
-		return nil, err
-	}
-	return grafanaDatasource, nil
-}
-
-// To migrate GrafanaDatasource from releaseNamespace to clusterNamespace for multi-tenancy,
-// we need to delete the old GrafanaDatasource in the releaseNamespace first.
-func (r *RegionalClusterRole) DeleteOldGrafanaDatasource() error {
-	grafanaDatasource := &grafanav1beta1.GrafanaDatasource{}
-	if err := r.client.Get(r.ctx, types.NamespacedName{
-		Name:      r.GetGrafanaDatasourceName(),
-		Namespace: r.releaseNamespace,
-	}, grafanaDatasource); err != nil {
-		if errors.IsNotFound(err) {
-			return nil
-		}
-		return err
-	}
-	if err := r.client.Delete(r.ctx, grafanaDatasource); err != nil {
-		return err
-	}
-	return nil
-}
-
-func (r *RegionalClusterRole) CreateOrUpdateGrafanaDatasource() error {
-	grafanaDatasource, err := r.GetGrafanaDatasource()
-	if err != nil {
-		return fmt.Errorf("failed to get grafana datasource: %v", err)
-	}
-
-	if grafanaDatasource == nil {
-		if err := r.DeleteOldGrafanaDatasource(); err != nil {
-			return fmt.Errorf("failed to delete old grafana datasource: %v", err)
-		}
-		if err := r.CreateGrafanaDataSource(); err != nil {
-			return fmt.Errorf("failed to create GrafanaDatasource: %v", err)
-		}
-		return nil
-	}
-
-	if err := r.UpdateGrafanaDatasource(grafanaDatasource); err != nil {
-		return fmt.Errorf("failed to update GrafanaDatasource: %v", err)
-	}
-
-	return nil
-}
-
-func (r *RegionalClusterRole) UpdateGrafanaDatasource(grafanaDatasource *grafanav1beta1.GrafanaDatasource) error {
-	logsEndpoint, err := getEndpoint(r.ctx, ReadLogsAnnotation, r.clusterDeployment, r.clusterDeploymentConfig)
-	if err != nil {
-		return err
-	}
-
-	httpClientConfig, err := r.GetHttpClientConfig()
-	if err != nil {
-		return fmt.Errorf("failed to get http client config: %v", err)
-	}
-
-	httpClientConfigRaw := r.httpClientConfigToRawJSON(httpClientConfig)
+func (r *RegionalClusterRole) UpdateConfigMap(cm *corev1.ConfigMap, configData *ConfigData) error {
+	configDataMap := configData.ToMap()
 
-	if logsEndpoint == grafanaDatasource.Spec.Datasource.URL &&
-		string(httpClientConfigRaw) == string(grafanaDatasource.Spec.Datasource.JSONData) {
+	if reflect.DeepEqual(cm.Data, configDataMap) {
```

Comment:
There is a better way:
```suggestion
	if maps.Equal(cm.Data, configDataMap) {
```
Please also replace the import.